### DISCUSSION TITLE
Regarding MakeCallback()

### DISCUSSION THREAD

**Discussion by deostroll** [nodejs/help#4859] [Q&A]
Which `MakeCallback()` implememtation is called here 👉https://github.com/nodejs/node/blob/52e4fb5b23157222d85b9275f7ccdfa88fe88d0f/src/node_file.cc#L581

Is it [this]( https://github.com/nodejs/node/blob/52e4fb5b23157222d85b9275f7ccdfa88fe88d0f/src/node_api.cc#L494)?

What about the case about the same call from [here](https://github.com/nodejs/node/blob/52e4fb5b23157222d85b9275f7ccdfa88fe88d0f/src/connection_wrap.cc#L73)?

Will it execute js code immediately? If so, how can I convince my self that it won't run concurrently when say two IO tasks simultaneously finished and wants to make a js callback?

Ps: I was tracing how `fs.readFile()` works under the hood.

**Comment by jasnell** [comment]
IO tasks do not run simultaneously. All of the IO callbacks for an isolate are executed one at a time.

**Reply by deostroll** [reply]
So what guarantees that callbacks in an isolate execute one at a time? Is it libuv?
